oidc-exchange.py

```diff
@@ -0,0 +1,156 @@
+
+import id
+import requests
+
+_GITHUB_STEP_SUMMARY = Path(os.getenv("GITHUB_STEP_SUMMARY"))
+
+_TOKEN_RETRIEVAL_FAILED_MESSAGE = dedent(
+    """
+    OIDC token retrieval failed: {identity_error}
+
+    This generally indicates a workflow configuration error, such as insufficient
+    permissions. Make sure that your workflow has `id-token: write` configured
+    at either the workflow or job level, e.g.:
+
+    ```yaml
+    permissions:
+      id-token: write
+    ```
+    """
+)
+
+
+def die(msg: str) -> NoReturn:
+    with _GITHUB_STEP_SUMMARY.open("a") as io:
+        print(msg, file=io)
+
+    # NOTE: `msg` is Markdown formatted, so we emit only the header line to
+    # avoid clogging the console log with a full Markdown formatted document.
+    print(f"::error::OIDC exchange failure: {msg.splitlines()[0]}", file=sys.stderr)
+    sys.exit(1)
+
+
+def debug(msg: str):
+    print(f"::debug::{msg}", file=sys.stderr)
+
+
+def get_normalized_input(name: str) -> str | None:
+    name = f"INPUT_{name.upper()}"
+    if val := os.getenv(name):
+        return val
+    return os.getenv(name.replace("-", "_"))
+
+
+def assert_successful_audience_call(resp: requests.Response, domain: str):
+    if resp.ok:
+        return
+
+    match resp.status_code:
+        case 403:
+            # This index supports OIDC, but forbids the client from using
+            # it (either because it's disabled, ratelimited, etc.)
+            die(f"audience retrieval failed: repository at {domain} has OIDC disabled")
+        case 404:
+            # This index does not support OIDC.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} does not indicate OIDC support"
+            )
+        case other:
+            # Unknown: the index may or may not support OIDC, but didn't respond with
+            # something we expect. This can happen if the index is broken, in maintenance mode,
+            # misconfigured, etc.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} responded with unexpected {other}"
+            )
+
+
+repository_url = get_normalized_input("repository-url")
+if not repository_url:
+    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
+    # hardcode the exchange endpoint and OIDC audience.
+    token_exchange_url = "https://pypi.org/_/oidc/github/mint-token"
+    oidc_audience = "pypi"
+else:
+    repository_domain = urlparse(repository_url).netloc
+    token_exchange_url = f"https://{repository_domain}/_/oidc/github/mint-token"
+
+    # Indices are expected to support `https://{domain}/_/oidc/audience`,
+    # which tells OIDC exchange clients which audience to use.
+    audience_url = f"https://{repository_domain}/_/oidc/audience"
+    audience_resp = requests.get(audience_url)
+    assert_successful_audience_call(audience_resp, repository_domain)
+
+    oidc_audience = audience_resp.json()["audience"]
```

Comment:
`oidc-exchange.py:76:1: WPS504 Found negated condition` is a dumb rule but...
```suggestion
if repository_url:
    repository_domain = urlparse(repository_url).netloc
    token_exchange_url = f"https://{repository_domain}/_/oidc/github/mint-token"

    # Indices are expected to support `https://{domain}/_/oidc/audience`,
    # which tells OIDC exchange clients which audience to use.
    audience_url = f"https://{repository_domain}/_/oidc/audience"
    audience_resp = requests.get(audience_url)
    assert_successful_audience_call(audience_resp, repository_domain)

    oidc_audience = audience_resp.json()["audience"]
else:
    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
    # hardcode the exchange endpoint and OIDC audience.
    token_exchange_url = "https://pypi.org/_/oidc/github/mint-token"
    oidc_audience = "pypi"
```

Reply:
This was a stylistic choice -- the negated case is the "happy" path, which is why it came first. I'm happy to accept this change if @webknjaz prefers it, though.

Reply:
It's totally acceptable to ignore linting rules where it makes sense, which is why the noqa is fine. OTOH, I'd not go as far as calling linting rules stupid without attempting to understand their original purpose.
In this case, a better solution wouldn't be inverting the if-else blocks but labeling the conditional clause by using an extra variable name.
Hence, having `is_pypi = not repository_url` followed by `if is_pypi:` would improve the situation by lowering the cognitive load a reader would experience.